File: src/wiki/lint/duplicate-detection.ts
// Duplicate page detection — programmatic candidate generation via shared links,
// bigram title similarity, and cross-language alias matching.
// Extracted from lint-fixes.ts to keep the module focused.

import { parseFrontmatter } from '../../core/frontmatter';
import {
  LINT_YIELD_EVERY_OUTER,
  LINT_YIELD_EVERY_PHASE1,
  LINT_YIELD_EVERY_COMPARISON,
} from '../../constants';

export interface DuplicateCandidate {
  target: string;
  source: string;
  reason: string;
  signal: 'crossLang' | 'bigram' | 'sharedLinks' | 'caseVariant';
  score: number;
}

// ── Pure Functions (extracted for testability) ───────────────────────────────

/** Extract character bigrams from string for similarity comparison. */
export function bigrams(s: string): Set<string> {
  const result = new Set<string>();
  const normalized = s.toLowerCase().replace(/[^a-z0-9一-鿿]/g, '');
  for (let i = 0; i < normalized.length - 1; i++) {
    result.add(normalized.substring(i, i + 2));
  }
  return result;
}

/** Normalize string for cross-language matching. */
export function normalizeForMatch(s: string): string {
  return s.toLowerCase().replace(/[\s\-_]+/g, '').replace(/[^a-z0-9一-鿿]/g, '');
}

const BODY_STOPWORDS = new Set([
  'also', 'are', 'been', 'being', 'both', 'but', 'can', 'could', 'did',
  'does', 'each', 'from', 'had', 'has', 'have', 'into', 'its', 'may',
  'might', 'must', 'not', 'only', 'other', 'our', 'shall', 'should',
  'than', 'that', 'the', 'their', 'them', 'then', 'there', 'these',
  'they', 'this', 'those', 'through', 'was', 'were', 'what', 'when',
  'where', 'which', 'while', 'will', 'with', 'would', 'your',
]);

/** Extract unique meaningful words from body text for content similarity comparison. */
export function bodyWordSet(text: string): Set<string> {
  return new Set(
    text.toLowerCase()
      .replace(/[^\w\s一-鿿]/g, ' ')
      .split(/\s+/)
      .filter(w => w.length > 3 && !BODY_STOPWORDS.has(w)),
  );
}

/** Compute Jaccard similarity between two sets. */
export function computeJaccard<T>(setA: Set<T>, setB: Set<T>): number {
  if (setA.size === 0 || setB.size === 0) return 0;
  let intersection = 0;
  for (const item of setA) {
    if (setB.has(item)) intersection++;
  }
  const union = setA.size + setB.size - intersection;
  return union > 0 ? intersection / union : 0;
}

// Generate duplicate-page candidates using programmatic signals.
// Returns candidates for LLM verification, capped by the O(n²) algorithm.
// Three signals, ordered by reliability:
//   1. Shared outgoing wiki-links (Jaccard >= 0.4)
//   2. Character bigram title similarity (catches spelling variants, same-language near-matches)
//   3. Cross-language alias match
export async function generateDuplicateCandidates(
  pages: Array<{ path: string; content: string; title: string }>,
): Promise<DuplicateCandidate[]> {
  interface PageMeta {
    path: string;
    title: string;
    aliases: string[];
    links: Set<string>;
    bodyWords: Set<string>;
  }

  const metas: PageMeta[] = [];
  const linkRegex = /\[\[([^\]|#]+)(?:[|#][^\]]+)?\]\]/g;

  for (let i = 0; i < pages.length; i++) {
    const page = pages[i];

    if (i > 0 && i % LINT_YIELD_EVERY_PHASE1 === 0) {
      await new Promise(resolve => window.setTimeout(resolve, 0));
    }

    const fm = parseFrontmatter(page.content);
    const aliases = Array.isArray(fm?.aliases) ? fm.aliases : [];

    const links = new Set<string>();
    const body = page.content.replace(/---[\s\S]*?---/, '');
    let match: RegExpExecArray | null;
    while ((match = linkRegex.exec(body)) !== null) {
      links.add(match[1].trim().toLowerCase());
    }

    // Strip wiki links before computing body words so link text doesn't inflate similarity
    const bodyText = body.replace(/\[\[[^\]]+\]\]/g, '');
    const bodyWords = bodyWordSet(bodyText);

    metas.push({ path: page.path, title: page.title, aliases, links, bodyWords });
  }

  const candidates = new Map<string, DuplicateCandidate>();

  const addCandidate = (pathA: string, pathB: string, reason: string, signal: DuplicateCandidate['signal'], score: number) => {
    const key = [pathA, pathB].sort().join('|||');
    if (!candidates.has(key)) {
      candidates.set(key, { target: pathA, source: pathB, reason, signal, score });
    } else if (score > candidates.get(key)!.score) {
      candidates.set(key, { target: pathA, source: pathB, reason, signal, score });
    }
  };

  let comparisonCount = 0;

  // Signal 1: Shared outgoing wiki-links (Jaccard >= 0.4)
  for (let i = 0; i < metas.length; i++) {
    if (i > 0 && i % LINT_YIELD_EVERY_OUTER === 0) {
      await new Promise(resolve => window.setTimeout(resolve, 0));
    }
    for (let j = i + 1; j < metas.length; j++) {
      comparisonCount++;
      if (comparisonCount % LINT_YIELD_EVERY_COMPARISON === 0) {
        await new Promise(resolve => window.setTimeout(resolve, 0));
      }

      const a = metas[i], b = metas[j];
      if (a.links.size === 0 || b.links.size === 0) continue;
      const jaccard = computeJaccard(a.links, b.links);
      if (jaccard >= 0.4) {
        // Body similarity gate: pages with different content are not duplicates
        // even if they share the same set of wiki-links (e.g., two unrelated pages
        // both linking only to one popular hub page).
        const bodySim = computeJaccard(a.bodyWords, b.bodyWords);
        if (bodySim < 0.2) continue;
        addCandidate(a.path, b.path, `Shared wiki-links (${Math.round(jaccard * 100)}% overlap)`, 'sharedLinks', jaccard);
      }
    }
  }

  // Signal 2: Bigram + cross-language on titles/aliases
  for (let i = 0; i < metas.length; i++) {
    if (i > 0 && i % LINT_YIELD_EVERY_OUTER === 0) {
      await new Promise(resolve => window.setTimeout(resolve, 0));
    }
    for (let j = i + 1; j < metas.length; j++) {
      comparisonCount++;
      if (comparisonCount % LINT_YIELD_EVERY_COMPARISON === 0) {
        await new Promise(resolve => window.setTimeout(resolve, 0));
      }

      const a = metas[i], b = metas[j];
      const namesA = [a.title, ...a.aliases];
      const namesB = [b.title, ...b.aliases];

      // 2a: Bigram similarity on all names (titles + aliases)
      let maxSim = 0;
      for (const nameA of namesA) {
        for (const nameB of namesB) {
          const sim = computeJaccard(bigrams(nameA), bigrams(nameB));
          if (sim > maxSim) maxSim = sim;
        }
      }
      if (maxSim >= 0.4) {
        addCandidate(a.path, b.path, `Title/alias similarity (${Math.round(maxSim * 100)}% match)`, 'bigram', maxSim);
      }

      // 2b: Cross-language alias match
      const normalizedNamesA = namesA.map(n => normalizeForMatch(n));
      const normalizedAliasesB = b.aliases.map(n => normalizeForMatch(n));
      const normalizedTitleB = normalizeForMatch(b.title);

      let crossLangMatch = false;
      for (const normA of normalizedNamesA) {
        if (normA && (normalizedAliasesB.includes(normA) || normalizedTitleB === normA)) {
          addCandidate(a.path, b.path, 'Cross-language match (alias or title overlap)', 'crossLang', 1.0);
          crossLangMatch = true;
          break;
        }
      }

      if (!crossLangMatch) {
        const normalizedNamesB = namesB.map(n => normalizeForMatch(n));
        const normalizedAliasesA = a.aliases.map(n => normalizeForMatch(n));
        const normalizedTitleA = normalizeForMatch(a.title);

        for (const normB of normalizedNamesB) {
          if (normB && (normalizedAliasesA.includes(normB) || normalizedTitleA === normB)) {
            addCandidate(a.path, b.path, 'Cross-language match (alias or title overlap)', 'crossLang', 1.0);
            break;
          }
        }
      }
    }
  }

  // Signal 3: Case-variant title collision
  // Two pages whose titles differ only in casing are highly likely duplicates.
  // e.g., "Unix" vs "unix", "Claude Code" vs "claude-code"
  for (let i = 0; i < metas.length; i++) {
    for (let j = i + 1; j < metas.length; j++) {
      const a = metas[i], b = metas[j];
      const lowerA = a.title.toLowerCase();
      const lowerB = b.title.toLowerCase();
      if (lowerA === lowerB && a.title !== b.title) {
        // Always pick lowercase-as-slug as target (deterministic merge direction)
        const [canonical, variant] = a.title < b.title ? [a, b] : [b, a];
        addCandidate(canonical.path, variant.path,
          `Case-variant duplicate: "${a.title}" ↔ "${b.title}"`, 'caseVariant', 0.9);
      }
    }
  }

  return Array.from(candidates.values());
}

File: python_lib/llm_wiki/wiki/lint/duplicate_detection.py
"""
Duplicate detection — port of src/wiki/lint/duplicate-detection.ts
"""
from __future__ import annotations

import re
from typing import Dict, List, Set

from ...core.frontmatter import parse_frontmatter


# ── Pure helpers ──────────────────────────────────────────────────────────────

def bigrams(s: str) -> Set[str]:
    normalized = re.sub(r"[^a-z0-9\u4e00-\u9fff]", "", s.lower())
    result: Set[str] = set()
    for i in range(len(normalized) - 1):
        result.add(normalized[i:i+2])
    return result


def normalize_for_match(s: str) -> str:
    return re.sub(r"[^a-z0-9\u4e00-\u9fff]", "", re.sub(r"[\s\-_]+", "", s.lower()))


_STOPWORDS = {
    "also", "are", "been", "being", "both", "but", "can", "could", "did",
    "does", "each", "from", "had", "has", "have", "into", "its", "may",
    "might", "must", "not", "only", "other", "our", "shall", "should",
    "than", "that", "the", "their", "them", "then", "there", "these",
    "they", "this", "those", "through", "was", "were", "what", "when",
    "where", "which", "while", "will", "with", "would", "your",
}


def body_word_set(text: str) -> Set[str]:
    words = re.sub(r"[^\w\s\u4e00-\u9fff]", " ", text.lower()).split()
    return {w for w in words if len(w) > 3 and w not in _STOPWORDS}


def compute_jaccard(set_a: Set, set_b: Set) -> float:
    if not set_a or not set_b:
        return 0.0
    intersection = len(set_a & set_b)
    union = len(set_a) + len(set_b) - intersection
    return intersection / union if union > 0 else 0.0


# ── Candidate generation ──────────────────────────────────────────────────────

class DuplicateCandidate:
    def __init__(self, target: str, source: str, reason: str, signal: str, score: float) -> None:
        self.target = target
        self.source = source
        self.reason = reason
        self.signal = signal  # 'crossLang' | 'bigram' | 'sharedLinks' | 'caseVariant'
        self.score = score


def generate_duplicate_candidates(
    pages: List[Dict],  # [{"path": str, "content": str, "title": str}]
) -> List[DuplicateCandidate]:
    """
    Programmatic duplicate candidate detection using three signals:
      1. Cross-language / alias overlap (crossLang)
      2. Bigram title similarity (bigram)
      3. Shared outgoing wiki-links (sharedLinks)
      4. Case-variant title collision (caseVariant)
    """
    link_re = re.compile(r"\[\[([^\]|#]+)(?:[|#][^\]]+)?\]\]")

    class PageMeta:
        def __init__(self, path: str, title: str, aliases: List[str],
                     links: Set[str], body_words: Set[str]) -> None:
            self.path = path
            self.title = title
            self.aliases = aliases
            self.links = links
            self.body_words = body_words

    metas: List[PageMeta] = []
    for page in pages:
        fm = parse_frontmatter(page["content"]) or {}
        aliases = fm.get("aliases", []) if isinstance(fm.get("aliases"), list) else []
        body = re.sub(r"---[\s\S]*?---", "", page["content"])
        links = {m.group(1).strip().lower() for m in link_re.finditer(body)}
        body_text = re.sub(r"\[\[[^\]]+\]\]", "", body)
        metas.append(PageMeta(page["path"], page["title"], aliases, links, body_word_set(body_text)))

    candidates: Dict[str, DuplicateCandidate] = {}

    def add(path_a: str, path_b: str, reason: str, signal: str, score: float) -> None:
        key = "|||".join(sorted([path_a, path_b]))
        if key not in candidates or score > candidates[key].score:
            candidates[key] = DuplicateCandidate(path_a, path_b, reason, signal, score)

    # Signal 1: Shared outgoing wiki-links (Jaccard >= 0.4, body sim >= 0.2)
    for i in range(len(metas)):
        for j in range(i + 1, len(metas)):
            a, b = metas[i], metas[j]
            if not a.links or not b.links:
                continue
            jaccard = compute_jaccard(a.links, b.links)
            if jaccard >= 0.4:
                body_sim = compute_jaccard(a.body_words, b.body_words)
                if body_sim >= 0.2:
                    add(a.path, b.path, f"Shared wiki-links ({round(jaccard * 100)}% overlap)", "sharedLinks", jaccard)

    # Signal 2: Bigram + cross-language on titles/aliases
    for i in range(len(metas)):
        for j in range(i + 1, len(metas)):
            a, b = metas[i], metas[j]
            names_a = [a.title] + a.aliases
            names_b = [b.title] + b.aliases

            # 2a: bigram similarity
            max_sim = 0.0
            for na in names_a:
                for nb in names_b:
                    sim = compute_jaccard(bigrams(na), bigrams(nb))
                    if sim > max_sim:
                        max_sim = sim
            if max_sim >= 0.4:
                add(a.path, b.path, f"Title/alias similarity ({round(max_sim * 100)}% match)", "bigram", max_sim)

            # 2b: cross-language normalized match
            norm_a = [normalize_for_match(n) for n in names_a]
            norm_b_aliases = [normalize_for_match(n) for n in b.aliases]
            norm_b_title = normalize_for_match(b.title)
            cross = False
            for na in norm_a:
                if na and (na in norm_b_aliases or na == norm_b_title):
                    add(a.path, b.path, "Cross-language match (alias or title overlap)", "crossLang", 1.0)
                    cross = True
                    break
            if not cross:
                norm_b = [normalize_for_match(n) for n in names_b]
                norm_a_aliases = [normalize_for_match(n) for n in a.aliases]
                norm_a_title = normalize_for_match(a.title)
                for nb in norm_b:
                    if nb and (nb in norm_a_aliases or nb == norm_a_title):
                        add(a.path, b.path, "Cross-language match (alias or title overlap)", "crossLang", 1.0)
                        break

    # Signal 3: Case-variant title collision
    for i in range(len(metas)):
        for j in range(i + 1, len(metas)):
            a, b = metas[i], metas[j]
            if a.title.lower() == b.title.lower() and a.title != b.title:
                canonical, variant = (a, b) if a.title < b.title else (b, a)
                add(canonical.path, variant.path,
                    f'Case-variant duplicate: "{a.title}" ↔ "{b.title}"',
                    "caseVariant", 0.9)

    return list(candidates.values())
